### PR TITLE
Solidity Upgrade + Gas Testing Infrastructure

### LINKED ISSUE #117: Investigate Upgrading Solidity Compiler Version
Considerations:

* Gas savings / optimizer improvements
* Were any bugfixes left off of current version?
* Security of newest version
* Contract changes required

### LINKED ISSUE #118: Investigate Removing SafeMath
Can we save gas and simplify the contracts by removing SafeMath and doing over/under flow protection inline where/if needed?

### PR DESCRIPTION
**Pull request by apbendi**
This PR includes the following changes:

* Add and configure the `hardhat-gas-reporter` plugin, which can be run with `yarn test:gas`
* Remove outdated "GasTester" contract
* Update the core Umbra contract to be compiled with solc 0.7.6 (closes #117)
* Remove SafeMath from the core Umbra contract (closes #118)
* Upgrade OpenZeppelin contracts to their latest versions

Here are some results from testing gas usage:

#### Initial Baseline

```
[1] ·---------------------------------------------|---------------------------|-------------|----------------------------·
[1] |            Solc version: 0.6.12             ·  Optimizer enabled: true  ·  Runs: 200  ·  Block limit: 6718946 gas  │
[1] ··············································|···························|·············|·····························
[1] |  Methods                                                                                                           │
[1] ·············|································|·············|·············|·············|··············|··············
[1] |  Contract  ·  Method                        ·  Min        ·  Max        ·  Avg        ·  # calls     ·  usd (avg)  │
[1] ·············|································|·············|·············|·············|··············|··············
[1] |  Umbra     ·  collectTolls                  ·          -  ·          -  ·      30431  ·           1  ·          -  │
[1] ·············|································|·············|·············|·············|··············|··············
[1] |  Umbra     ·  sendEth                       ·          -  ·          -  ·      35404  ·          11  ·          -  │
[1] ·············|································|·············|·············|·············|··············|··············
[1] |  Umbra     ·  sendToken                     ·      43310  ·      73322  ·      57719  ·          25  ·          -  │
[1] ·············|································|·············|·············|·············|··············|··············
[1] |  Umbra     ·  setToll                       ·          -  ·          -  ·      27494  ·           6  ·          -  │
[1] ·············|································|·············|·············|·············|··············|··············
[1] |  Umbra     ·  withdrawToken                 ·          -  ·          -  ·      47876  ·           3  ·          -  │
[1] ·············|································|·············|·············|·············|··············|··············
[1] |  Umbra     ·  withdrawTokenAndCall          ·      33768  ·     143058  ·     115636  ·           8  ·          -  │
[1] ·············|································|·············|·············|·············|··············|··············
[1] |  Umbra     ·  withdrawTokenAndCallOnBehalf  ·      69556  ·     217234  ·     180161  ·           8  ·          -  │
[1] ·············|································|·············|·············|·············|··············|··············
[1] |  Umbra     ·  withdrawTokenOnBehalf         ·          -  ·          -  ·      83394  ·           3  ·          -  │
[1] ·············|································|·············|·············|·············|··············|··············
[1] |  Deployments                                ·                                         ·  % of limit  ·             │
[1] ··············································|·············|·············|·············|··············|··············
[1] |  Umbra                                      ·          -  ·          -  ·    1463924  ·      21.8 %  ·          -  │
[1] ·---------------------------------------------|-------------|-------------|-------------|--------------|-------------·
```

#### Go from 200 Optimizer Runs to 10K

```
[1] ·---------------------------------------------|---------------------------|---------------|----------------------------·
[1] |            Solc version: 0.6.12             ·  Optimizer enabled: true  ·  Runs: 10000  ·  Block limit: 6718946 gas  │
[1] ··············································|···························|···············|·····························
[1] |  Methods                                                                                                             │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Contract  ·  Method                        ·  Min        ·  Max        ·  Avg          ·  # calls     ·  usd (avg)  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  collectTolls                  ·          -  ·          -  ·        30431  ·           1  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  sendEth                       ·          -  ·          -  ·        35382  ·          11  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  sendToken                     ·      43142  ·      73142  ·        57540  ·          25  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  setToll                       ·          -  ·          -  ·        27472  ·           6  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawToken                 ·          -  ·          -  ·        47759  ·           3  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawTokenAndCall          ·      33628  ·     142879  ·       115467  ·           8  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawTokenAndCallOnBehalf  ·      69267  ·     216906  ·       179843  ·           8  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawTokenOnBehalf         ·          -  ·          -  ·        83140  ·           3  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Deployments                                ·                                           ·  % of limit  ·             │
[1] ··············································|·············|·············|···············|··············|··············
[1] |  Umbra                                      ·          -  ·          -  ·      1550220  ·      23.1 %  ·          -  │
[1] ·---------------------------------------------|-------------|-------------|---------------|--------------|-------------·
```

#### Upgrade to solc 0.7.6

```
[1] ·---------------------------------------------|---------------------------|---------------|----------------------------·
[1] |             Solc version: 0.7.6             ·  Optimizer enabled: true  ·  Runs: 10000  ·  Block limit: 6718946 gas  │
[1] ··············································|···························|···············|·····························
[1] |  Methods                                                                                                             │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Contract  ·  Method                        ·  Min        ·  Max        ·  Avg          ·  # calls     ·  usd (avg)  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  collectTolls                  ·          -  ·          -  ·        30431  ·           1  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  sendEth                       ·          -  ·          -  ·        35382  ·          11  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  sendToken                     ·      43142  ·      73142  ·        57540  ·          25  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  setToll                       ·          -  ·          -  ·        27472  ·           6  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawToken                 ·          -  ·          -  ·        47759  ·           3  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawTokenAndCall          ·      33628  ·     142896  ·       110668  ·           8  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawTokenAndCallOnBehalf  ·      69267  ·     216923  ·       175044  ·           8  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawTokenOnBehalf         ·          -  ·          -  ·        83128  ·           3  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Deployments                                ·                                           ·  % of limit  ·             │
[1] ··············································|·············|·············|···············|··············|··············
[1] |  Umbra                                      ·          -  ·          -  ·      1549992  ·      23.1 %  ·          -  │
[1] ·---------------------------------------------|-------------|-------------|---------------|--------------|-------------·
```

#### Remove SafeMath

```
[1] ·---------------------------------------------|---------------------------|---------------|----------------------------·
[1] |             Solc version: 0.7.6             ·  Optimizer enabled: true  ·  Runs: 10000  ·  Block limit: 6718946 gas  │
[1] ··············································|···························|···············|·····························
[1] |  Methods                                                                                                             │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Contract  ·  Method                        ·  Min        ·  Max        ·  Avg          ·  # calls     ·  usd (avg)  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  collectTolls                  ·          -  ·          -  ·        30431  ·           1  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  sendEth                       ·          -  ·          -  ·        35310  ·          11  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  sendToken                     ·      43142  ·      73142  ·        57540  ·          25  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  setToll                       ·          -  ·          -  ·        27472  ·           6  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawToken                 ·          -  ·          -  ·        47759  ·           3  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawTokenAndCall          ·      33628  ·     142896  ·       110668  ·           8  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawTokenAndCallOnBehalf  ·      69267  ·     216911  ·       175038  ·           8  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawTokenOnBehalf         ·          -  ·          -  ·        83128  ·           3  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Deployments                                ·                                           ·  % of limit  ·             │
[1] ··············································|·············|·············|···············|··············|··············
[1] |  Umbra                                      ·          -  ·          -  ·      1526492  ·      22.7 %  ·          -  │
[1] ·---------------------------------------------|-------------|-------------|---------------|--------------|-------------·
```

#### Upgrade OZ Contracts

```
[1] ·---------------------------------------------|---------------------------|---------------|----------------------------·
[1] |             Solc version: 0.7.6             ·  Optimizer enabled: true  ·  Runs: 10000  ·  Block limit: 6718946 gas  │
[1] ··············································|···························|···············|·····························
[1] |  Methods                                                                                                             │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Contract  ·  Method                        ·  Min        ·  Max        ·  Avg          ·  # calls     ·  usd (avg)  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  collectTolls                  ·          -  ·          -  ·        30431  ·           1  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  sendEth                       ·          -  ·          -  ·        35310  ·          11  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  sendToken                     ·      43142  ·      73142  ·        57540  ·          25  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  setToll                       ·          -  ·          -  ·        27472  ·           6  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawToken                 ·          -  ·          -  ·        47759  ·           3  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawTokenAndCall          ·      33628  ·     142896  ·       110668  ·           8  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawTokenAndCallOnBehalf  ·      69255  ·     216923  ·       175038  ·           8  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Umbra     ·  withdrawTokenOnBehalf         ·          -  ·          -  ·        83140  ·           3  ·          -  │
[1] ·············|································|·············|·············|···············|··············|··············
[1] |  Deployments                                ·                                           ·  % of limit  ·             │
[1] ··············································|·············|·············|···············|··············|··············
[1] |  Umbra                                      ·          -  ·          -  ·      1526492  ·      22.7 %  ·          -  │
[1] ·---------------------------------------------|-------------|-------------|---------------|--------------|-------------·
```

### CODE REVIEW

File: contracts/hardhat.config.ts [hunk 1]
```diff
@@ -80,19 +84,39 @@ const config: HardhatUserConfig = {
...
+        version: '0.6.12',
+        settings: {
+          // https://hardhat.org/hardhat-network/#solidity-optimizer-support
+          optimizer: {
+            enabled: true,
+            runs: 10000,
+          },
+        },
+      },
+    ],
```
Comment (mds1) [inline contracts/hardhat.config.ts:108]: Small gas savings if we bump the runs to 999,999. If we bump to 9,999,999 it doesn't really make a difference, so I'd suggest we use 999,999 runs for both solc versions

File: contracts/package.json [hunk 2]
```diff
@@ -17,8 +15,8 @@
     "@nomiclabs/hardhat-truffle5": "^2.0.0",
     "@nomiclabs/hardhat-waffle": "^2.0.0",
     "@nomiclabs/hardhat-web3": "^2.0.0",
-    "@openzeppelin/contracts": "^3.2.0",
     "@openzeppelin/test-helpers": "^0.5.10",
+    "@truffle/debug-utils": "^4.1.1",
```
Comment (mds1) [inline contracts/package.json:19]: Just curious what this package is for?

**Review by mds1 — CHANGES_REQUESTED** [review]
Looks great and works great! One suggestion about changing the number of optimizer runs

File: contracts/package.json (hunk 2 above)
Comment (apbendi) [reply contracts/package.json:19]: I'm not sure, we already had it. I just moved it from dependencies to dev dependencies.

**Pull request merged**